File: Darts/src/burlton/dartzee/code/bean/PlayerAvatar.java
package burlton.dartzee.code.bean;

import burlton.dartzee.code.db.PlayerEntity;
import burlton.dartzee.code.db.PlayerImageEntity;
import burlton.dartzee.code.screen.PlayerImageDialog;

import javax.swing.*;
import javax.swing.border.EtchedBorder;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class PlayerAvatar extends JLabel
{
	private static final ImageIcon AVATAR_UNSET = new ImageIcon(PlayerAvatar.class.getResource("/avatars/Unset.png"));
	
	private PlayerEntity player = null;
	private String avatarId = "";
	private boolean readOnly = false;
	
	public PlayerAvatar()
	{
		super(AVATAR_UNSET);
		setPreferredSize(new Dimension(150, 150));
		setBorder(new EtchedBorder(EtchedBorder.RAISED, null, null));
		setHorizontalAlignment(SwingConstants.CENTER);
		
		addMouseListener(new AvatarClickListener());
	}
	
	public void init(PlayerEntity player, boolean saveChanges)
	{
		//Only set the player variable if we want to allow the label to directly make changes to it.
		if (saveChanges)
		{
			this.player = player;
		}
		
		if (player != null)
		{
			this.avatarId = player.getPlayerImageId();
			
			ImageIcon newIcon = player.getAvatar();
			setIcon(newIcon);
		}
		else
		{
			avatarId = "";
			setIcon(AVATAR_UNSET);
		}
	}
	
	/**
	 * Gets / Sets
	 */
	public String getAvatarId()
	{
		return avatarId;
	}
	public void setReadOnly(boolean readOnly)
	{
		this.readOnly = readOnly;
	}
	
	/**
	 * MouseListener
	 */
	private final class AvatarClickListener extends MouseAdapter
	{
		@Override
		public void mouseClicked(MouseEvent arg0)
		{
			if (readOnly)
			{
				return;
			}
			
			PlayerImageDialog dlg = new PlayerImageDialog();
			dlg.setVisible(true);
			
			avatarId = dlg.getPlayerImageIdSelected();
			if (!avatarId.isEmpty())
			{
				ImageIcon newIcon = PlayerImageEntity.retrieveImageIconForId(avatarId);
				setIcon(newIcon);
				
				if (player != null)
				{
					player.setPlayerImageId(avatarId);
					player.saveToDatabase();
				}
			}
		}
		
		@Override
		public void mouseEntered(MouseEvent arg0)
		{
			if (!readOnly)
			{
				setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
			}
		}
		
		@Override
		public void mouseExited(MouseEvent arg0)
		{
			if (!readOnly)
			{
				setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
			}
		}
	}
}

File: Darts/src/burlton/dartzee/code/bean/PlayerAvatar.kt
package burlton.dartzee.code.bean

import burlton.dartzee.code.db.PlayerEntity
import burlton.dartzee.code.db.PlayerImageEntity
import burlton.dartzee.code.screen.PlayerImageDialog
import java.awt.Color
import java.awt.Cursor
import java.awt.Dimension
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.ImageIcon
import javax.swing.JLabel
import javax.swing.SwingConstants
import javax.swing.border.EtchedBorder
import javax.swing.border.LineBorder

class PlayerAvatar : JLabel(AVATAR_UNSET)
{
    private var player: PlayerEntity? = null

    var avatarId = ""
    var readOnly = false

    init
    {
        preferredSize = Dimension(150, 150)
        border = EtchedBorder(EtchedBorder.RAISED, null, null)
        horizontalAlignment = SwingConstants.CENTER

        addMouseListener(AvatarClickListener())
    }

    fun setSelected(selected: Boolean)
    {
        border = if (selected) LineBorder(Color.RED, 2) else EtchedBorder(EtchedBorder.RAISED, null, null)
    }

    fun init(player: PlayerEntity?, saveChanges: Boolean)
    {
        //Only set the player variable if we want to allow the label to directly make changes to it.
        if (saveChanges)
        {
            this.player = player
        }

        avatarId = player?.playerImageId ?: ""
        icon = player?.getAvatar() ?: AVATAR_UNSET
    }

    /**
     * MouseListener
     */
    private inner class AvatarClickListener : MouseAdapter()
    {
        override fun mouseClicked(arg0: MouseEvent?)
        {
            if (readOnly)
            {
                return
            }

            val dlg = PlayerImageDialog()
            dlg.isVisible = true

            avatarId = dlg.playerImageIdSelected

            if (!avatarId.isEmpty())
            {
                val newIcon = PlayerImageEntity.retrieveImageIconForId(avatarId)
                icon = newIcon

                player?.let{
                    it.playerImageId = avatarId
                    it.saveToDatabase()
                }
            }
        }

        override fun mouseEntered(arg0: MouseEvent?)
        {
            if (!readOnly)
            {
                cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            }
        }

        override fun mouseExited(arg0: MouseEvent?)
        {
            if (!readOnly)
            {
                cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
            }
        }
    }

    companion object
    {
        private val AVATAR_UNSET = ImageIcon(PlayerAvatar::class.java.getResource("/avatars/Unset.png"))
    }
}
